File: ch7/7-2.go
package main

import (
	"bufio"
	"fmt"
	"os"
)

const (
	linelen = 10
	octLen  = 3
)

func main() {
	r := bufio.NewReader(os.Stdin)
	for {
		b, err := r.ReadByte()
		if err != nil {
			break
		}
		var cnum int
		if int(b) < 32 {
			cnum = printoct(int(b), cnum)
		} else {
			cnum = printnor(int(b), cnum)
		}
		if cnum >= linelen {
			cnum = cnum - linelen
		}
	}
}

func printoct(c, n int) int {
	var l int
	l = n + octLen
	if l <= linelen {
		fmt.Printf("%o", c)
	} else {
		fmt.Printf("\n%o", c)
	}
	return l
}

func printnor(c, n int) int {
	var l int
	l = n + 1
	if l <= linelen {
		fmt.Printf("%c", c)
	} else {
		fmt.Printf("\n%c", c)
	}
	return l
}

File: ch7/7-2.c
#include <stdio.h>

#define LINELEN 10
#define OCTLEN 3

int printoct(int c, int n);
int printnor(int c, int n);

int main(void)
{
    int c, cnum = 0;
    while ((c = getchar()) != EOF) {
        cnum = (c < 32) ? printoct(c, cnum) : printnor(c, cnum);
        if (cnum >= LINELEN) {
            cnum = cnum - LINELEN;
        }
    }
    return 0;
}

int printoct(int c, int n)
{
    int len;
    if ((len = (n + OCTLEN)) <= LINELEN) {
        printf("%o", c);
    } else {
        printf("\n%o", c);
    }
    return len;
}

int printnor(int c, int n)
{
    int len;
    if ((len = (n + 1)) <= LINELEN) {
        printf("%c", c);
    } else {
        printf("\n%c", c);
    }
    return len;
}
